// @shopify/cli-kit/dist/private/node/ui/components/TokenizedText.js
import { Command } from './Command.js';
import { Link } from './Link.js';
import { List } from './List.js';
import { UserInput } from './UserInput.js';
import { FilePath } from './FilePath.js';
import { Subdued } from './Subdued.js';
import { Box, Text } from 'ink';
import React from 'react';
function tokenToBlock(token) {
    return {
        display: typeof token !== 'string' && 'list' in token ? 'block' : 'inline',
        value: token,
    };
}
export function tokenItemToString(token) {
    if (typeof token === 'string') {
        return token;
    }
    else if ('command' in token) {
        return token.command;
    }
    else if ('link' in token) {
        return token.link.label || token.link.url;
    }
    else if ('char' in token) {
        return token.char;
    }
    else if ('userInput' in token) {
        return token.userInput;
    }
    else if ('subdued' in token) {
        return token.subdued;
    }
    else if ('filePath' in token) {
        return token.filePath;
    }
    else if ('list' in token) {
        return token.list.items.map(tokenItemToString).join(' ');
    }
    else {
        return token.map(tokenItemToString).join(' ');
    }
}
export function appendToTokenItem(token, suffix) {
    return Array.isArray(token) ? [...token, suffix] : [token, suffix];
}
function splitByDisplayType(acc, item) {
    if (item.display === 'block') {
        acc.push([item]);
    }
    else {
        const last = acc[acc.length - 1];
        if (last && last[0].display === 'inline') {
            last.push(item);
        }
        else {
            acc.push([item]);
        }
    }
    return acc;
}
/**
 * `TokenizedText` renders a text string with tokens that can be either strings,
 * links, and commands.
 */
const TokenizedText = ({ item }) => {
    if (typeof item === 'string') {
        return React.createElement(Text, null, item);
    }
    else if ('command' in item) {
        return React.createElement(Command, { command: item.command });
    }
    else if ('link' in item) {
        return React.createElement(Link, { ...item.link });
    }
    else if ('char' in item) {
        return React.createElement(Text, null, item.char[0]);
    }
    else if ('userInput' in item) {
        return React.createElement(UserInput, { userInput: item.userInput });
    }
    else if ('subdued' in item) {
        return React.createElement(Subdued, { subdued: item.subdued });
    }
    else if ('filePath' in item) {
        return React.createElement(FilePath, { filePath: item.filePath });
    }
    else if ('list' in item) {
        return React.createElement(List, { ...item.list });
    }
    else {
        const groupedItems = item.map(tokenToBlock).reduce(splitByDisplayType, []);
        return (React.createElement(Box, { flexDirection: "column" }, groupedItems.map((items, groupIndex) => {
            if (items[0].display === 'inline') {
                return (React.createElement(Text, { key: groupIndex }, items.map((item, itemIndex) => (React.createElement(Text, { key: itemIndex },
                    itemIndex !== 0 && !(typeof item.value !== 'string' && 'char' in item.value) && React.createElement(Text, null, " "),
                    React.createElement(TokenizedText, { item: item.value }))))));
            }
            else {
                return React.createElement(List, { key: groupIndex, ...items[0].value.list });
            }
        })));
    }
};
export { TokenizedText };
